payload/src/main/java/io/jenkins/jenkinsfile/runner/Runner.java

```diff
@@ -53,9 +57,18 @@ public int run(Bootstrap bootstrap) throws Exception {
           w.setDefinition(new PipelineAsYamlScriptFlowDefinition(
             FileUtils.readFileToString(bootstrap.jenkinsfile),!bootstrap.noSandBox));
         } else {
-          w.setDefinition(new CpsScmFlowDefinition(
-            new FileSystemSCM(bootstrap.jenkinsfile.getParent()), bootstrap.jenkinsfile.getName()));
-          workflowActionsList.add(new SetJenkinsfileLocation(bootstrap.jenkinsfile, !bootstrap.noSandBox));
+            if (bootstrap.scm != null) {
+                SCMContainer scm = SCMContainer.loadFromYAML(bootstrap.scm);
+                Credentials fromSCM = scm.getCredential();
+                if (fromSCM != null) {
+                    addCredentials(fromSCM);
+                }
+                w.setDefinition(new CpsScmFlowDefinition(scm.getSCM(), bootstrap.jenkinsfile.getName()));
+            } else {
+                w.setDefinition(new CpsScmFlowDefinition(
+                        new FileSystemSCM(bootstrap.jenkinsfile.getParent()), bootstrap.jenkinsfile.getName()));
+            }
+            workflowActionsList.add(new SetJenkinsfileLocation(bootstrap.jenkinsfile, !bootstrap.noSandBox));
```

Comment:
So I think in a perfect world, we'd actually just use the Jenkinsfile from the SCM when using an SCM, rather than the one on disk, but that would mean reimplementing [all of this](https://github.com/jenkinsci/workflow-cps-plugin/blob/master/src/main/java/org/jenkinsci/plugins/workflow/cps/CpsScmFlowDefinition.java#L102-L192), in order to still allow disabling the sandbox. Which is not worth it. So...

Reply:
Yes. Patch in the upstream may make sense, at least on the API level